umodbus/async/serial.py

```diff
@@ -0,0 +1,304 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2019, Pycom Limited.
+#
+# This software is licensed under the GNU GPL version 3 or any
+# later version, with permitted additional terms. For more information
+# see the Pycom Licence v1.0 document supplied with this file, or
+# available at https://www.pycom.io/opensource/licensing
+#
+
+# system packages
+from ..sys_imports import struct
+from ..sys_imports import List, Literal, Optional, Tuple, Union
+
+try:
+    import uasyncio as asyncio
+    from machine import UART, Pin
+except ImportError:
+    import asyncio
+
+# custom packages
+from .. import functions, const as Const
+from .common import AsyncRequest
+from .modbus import Modbus
+
+
+class SerialServer(AbstractServerInterface):
+    def __init__(self,
+                 uart_id:int = 1,
+                 baudrate:int = 9600,
+                 data_bits:int = 8,
+                 stop_bits:int = 1,
+                 parity=None,
+                 pins: Tuple[int, int] = (13, 14),
+                 ctrl_pin=None):
+        super().__init__()
+        self._uart = UART(uart_id,
+                          baudrate=baudrate,
+                          bits=data_bits,
+                          parity=parity,
+                          stop=stop_bits,
+                          # timeout_chars=2,  # WiPy only
+                          # pins=pins         # WiPy only
+                          tx=pins[0],
+                          rx=pins[1]
+                          )
+        self.uart_in = asyncio.StreamReader(self._uart)
+        self.uart_out = asyncio.StreamWriter(self._uart, {}, reader=self.uart_in, loop=asyncio.get_running_loop())
+
+        if baudrate <= 19200:
+            # 4010us (approx. 4ms) @ 9600 baud
+            self._t35chars = (3500000 * (data_bits + stop_bits + 2)) // baudrate
+        else:
+            self._t35chars = 1750   # 1750us (approx. 1.75ms)
+
+        if ctrl_pin is not None:
+            self._ctrlPin = Pin(ctrl_pin, mode=Pin.OUT)
+        else:
+            self._ctrlPin = None
+
+    def bind(self, local_ip: str, local_port: int = 502, max_connections: int = 10) -> None:
```

Comment:
doc strings are missing for this file and there might be some flake8 warnings ;)

Reply:
I will probably have to overhaul this class to actually make it functional, so a lot of functions will be changed in the meantime; I'll add doc strings after that's complete.